python/ray/dag/compiled_dag_node.py

```diff
@@ -375,61 +410,103 @@ def prepare(self):
+                        assert self._recv_stream == nccl_group.recv_stream, (
+                            "Currenlty all torch tensor input channels of a "
+                            "Compiled Graph task should use the same recv cuda stream."
+                        )
+                    self._recv_stream = nccl_group.recv_stream
+
+    def wrap_and_set_intermediate_future(
+        self, val: Any, wrap_in_gpu_future: bool
+    ) -> None:
         """
-        Store the intermediate result of a READ or COMPUTE operation.
+        Wrap the value in a `DAGOperationFuture` and store to the intermediate future.
+        The value corresponds to result of a READ or COMPUTE operation.
+
+        If wrap_in_gpu_future is True, the value will be wrapped in a _GPUFuture,
+        Otherwise, the future will be a ResolvedFuture.
 
         Args:
-            data: The intermediate result of a READ or COMPUTE operation.
+            val: The value to wrap in a future.
+            wrap_in_gpu_future: Whether to wrap the value in a _GPUFuture.
         """
-        assert self._intermediate_buffer is None
-        self._intermediate_buffer = data
+        assert self._intermediate_future is None
+        import cupy as cp
 
-    def reset_intermediate_buffer(self) -> Any:
-        """
-        Retrieve the intermediate result of a READ or COMPUTE operation,
-        and reset the intermediate buffer to None.
+        if wrap_in_gpu_future:
+            future = GPUFuture(val, cp.cuda.get_current_stream())
+        else:
+            future = ResolvedFuture(val)
+        self._intermediate_future = future
 
+    def reset_and_wait_intermediate_future(self) -> Any:
+        """
+        Reset the intermediate future and wait for the result.
         Returns:
-            The intermediate result of a READ or COMPUTE operation.
+            The result of a READ or COMPUTE operation from the intermediate future.
         """
-        data = self._intermediate_buffer
-        self._intermediate_buffer = None
-        return data
+        future = self._intermediate_future
+        self._intermediate_future = None
+        return future.wait()
 
-    def _read(self) -> bool:
+    def _read(self, overlap_gpu_communication: bool) -> bool:
         """
         Read input data from upstream DAG nodes and cache the intermediate result.
 
+        Args:
+            overlap_gpu_communication: Whether to overlap GPU communication with
+                computation during DAG execution to improve performance.
+
         Returns:
             True if system error occurs and exit the loop; otherwise, False.
         """
-        assert self._intermediate_buffer is None
+        assert self._intermediate_future is None
         exit = False
         try:
             input_data = self.input_reader.read()
-            self.set_intermediate_buffer(input_data)
+            self.wrap_and_set_intermediate_future(input_data, overlap_gpu_communication)
```

Comment:
Might also want to add comments here and elsewhere explaining why we want to wrap in a GPU future if overlapping is enabled.

```suggestion
            self.wrap_and_set_intermediate_future(input_data, wrap_in_gpu_future=overlap_gpu_communication)
```